File: libraries/leo2-common/src/test/java/org/deku/leo2/messaging/LogTest.java
package org.deku.leo2.messaging;

import ch.qos.logback.classic.Logger;
import org.deku.leo2.MessagingTest;
import org.deku.leo2.messaging.activemq.ActiveMQContext;
import org.deku.leo2.messaging.log.LogAppender;
import org.deku.leo2.messaging.log.LogListener;
import org.junit.Ignore;
import org.junit.Test;
import org.slf4j.LoggerFactory;

import javax.jms.JMSException;

/**
 * Created by masc on 16.04.15.
 */
@Ignore
public class LogTest extends MessagingTest {
    @Test
    public void testSend() throws JMSException {
        // Setup log appender
        LogAppender lAppender = new LogAppender(ActiveMQContext.instance());
        lAppender.start();
        Logger lRoot = (Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);
        lRoot.addAppender(lAppender);

        // Generate some log messages
        for (int i = 0; i < 100; i++)
            lRoot.info("Test!");

        // Dispose to make sure everything is flushed
        lAppender.dispose();
    }

    @Test
    public void testReceive() throws JMSException, InterruptedException {
        // Setup log message listener
        LogListener mListener = new LogListener(ActiveMQContext.instance());
        mListener.start();

        // Wait for some messages to be received
        Thread.sleep(20000);

        mListener.dispose();
    }
}

File: leoz-core/src/prototype/java/org/deku/leoz/log/LogTest.kt
package org.deku.leoz.log

import ch.qos.logback.classic.Logger
import com.github.salomonbrys.kodein.Kodein
import com.github.salomonbrys.kodein.conf.global
import com.github.salomonbrys.kodein.erased.instance
import com.github.salomonbrys.kodein.lazy
import org.deku.leoz.SystemInformation
import org.deku.leoz.bundle.BundleType
import org.deku.leoz.config.JmsConfiguration
import org.deku.leoz.config.JmsEndpoints
import org.deku.leoz.config.MessagingTestConfiguration
import org.deku.leoz.identity.DesktopIdentityFactory
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.slf4j.LoggerFactory
import sx.mq.MqChannel
import sx.mq.MqHandler
import sx.mq.jms.activemq.ActiveMQBroker
import sx.mq.jms.channel
import sx.mq.jms.listeners.SpringJmsListener

import javax.jms.JMSException
import java.util.concurrent.Executors

/**
 * @author masc
 */
class LogTest {
    private val log = LoggerFactory.getLogger(this.javaClass)

    init {
        Kodein.global.addImport(MessagingTestConfiguration.module)
    }

    val broker: ActiveMQBroker by Kodein.global.lazy.instance()

    @Before
    fun setup() {
        this.broker.start()
    }

    @After
    fun tearDown() {
        this.broker.stop()
    }

    @Test
    @Throws(JMSException::class)
    fun testSend() {
        // Setup log appender
        val logAppender = LogMqAppender(
                channelSupplier = { JmsEndpoints.central.transient.kryo.channel() },
                identitySupplier = {
                    DesktopIdentityFactory(BundleType.LEOZ_NODE.value, SystemInformation.Companion.create()).create()
                })
        logAppender.start()

        val logRoot = LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME) as Logger
        logRoot.addAppender(logAppender)

        // Generate some log messages
        for (i in 0..10000)
            logRoot.info("Test!")

        // Dispose to make sure everything is flushed
        logAppender.flush()
        logAppender.close()
    }

    @Test
    @Throws(JMSException::class, InterruptedException::class)
    fun testReceive() {
        // Setup log message listener
        val listener = object : SpringJmsListener(
                JmsEndpoints.central.transient.kryo,
                Executors.newSingleThreadExecutor()) {

        }

        listener.addDelegate(object : MqHandler<LogMessage> {
            override fun onMessage(message: LogMessage, replyChannel: MqChannel?) {
                log.info("${message}: ${message.logEntries.count()}")
            }
        })

        listener.start()

        // Wait for some messages to be received
        Thread.sleep(20000)

        listener.close()
    }
}

